### DISCUSSION TITLE
How do I actually practically connect to a vsock?

### DISCUSSION THREAD

**Discussion by kanpov** [firecracker-microvm/firecracker#4686] [Q&A] [answered]
In the vsock example, the way specified is via socat and then issuing the CONNECT request to a given port, for example 8000. How do I actually do this anywhere except for socat?

8000, the port is the request url but it's not a valid url so anything except for socat seems to reject it. Curl rejects 8000 as a uri, reqwest, hyper etc. all reject 8000 or any other port because it's an invalid url. Then how am I supposed to actually connect?

The whole "port request" mechanism with CONNECT seems exceptionally inconvenient, I haven't found a way to even replicate it outside of socat which is unusable for any sort of automation.

**Comment by kanpov** [comment]
The only idea I currently have is to instantiate the socat command, write CONNECT to its stdin, then implement hyper's I/O traits so that they redirect the writes and reads to and from the ChildStdin/ChildStdout of the socat process. But obviously, this is a horrible idea and there must be some sort of actual way to use this feature.

**Comment by kanpov** [comment, marked as answer]
I figured out how to do this: I basically made my own tower service where I do this when creating a connection:

1. Create a `tokio::net::UnixStream` like normal
2. Write the plaintext CONNECT command to the stream and flush it
3. Read the next line (OK ...) to determine if the tunnel was established and move past the CONNECT and OK lines
4. Pass the `UnixStream` further so that hyper can use it

I'm currently making a `hyper-sockets` crate where I'll share my implementation.

Even so, this is remarkably hard to do!

**Comment by kalyazin** [comment]
Hi @kanpov ! Glad to hear that you have found the solution. That is very similar to what we do.

**Reply by kanpov** [reply]
> Hi @kanpov ! Glad to hear that you have found the solution. That is very similar to what we do.

Afterwards I also made a crate (https://crates.io/crates/hyper-client-sockets) that has support for the entire Firecracker "vsock stack", so connecting to unix, vsock, and these "firecracker-flavored" unix sockets. It might be useful for others that use Rust and Firecracker.
